lib/fluent/plugin/out_forward.rb

```diff
@@ -1079,38 +1065,54 @@ def on_read(sock, ri, data)
 
       private
 
-      def connect(host = nil)
-        socket, request_info =
-                if @keepalive
-                  ri = RequestInfo.new(:established)
-                  sock = @socket_cache.fetch_or do
-                    s = @sender.create_transfer_socket(host || resolved_host, port, @hostname)
-                    ri = RequestInfo.new(@sender.security ? :helo : :established) # overwrite if new connection
-                    s
-                  end
-                  [sock, ri]
-                else
-                  @log.debug('connect new socket')
-                  [@sender.create_transfer_socket(host || resolved_host, port, @hostname), RequestInfo.new(@sender.security ? :helo : :established)]
-                end
+      def connect(host = nil, require_ack: false, &block)
+        if @keepalive
+          return connect_keepalive(host, require_ack: require_ack, &block)
+        end
+
+        @log.debug('connect new socket')
+        socket = @sender.create_transfer_socket(host || resolved_host, port, @hostname)
+        request_info = RequestInfo.new(@sender.security ? :helo : :established)
+
+        unless block_given?
+          return [socket, request_info]
+        end
+
+        begin
+          yield(socket, request_info)
+        ensure
+          unless require_ack
+            socket.close_write rescue nil
+            socket.close rescue nil
+          end
+        end
+      end
 
-        if block_given?
-          ret = nil
+      def connect_keepalive(host = nil, require_ack: false)
+        request_info = RequestInfo.new(:established)
+        socket = @socket_cache.fetch_or do
+          s = @sender.create_transfer_socket(host || resolved_host, port, @hostname)
+          request_info = RequestInfo.new(@sender.security ? :helo : :established) # overwrite if new connection
+          s
+        end
+
+        unless block_given?
+          return [socket, request_info]
+        end
+
+        ret =
```

Comment:
This `ret` seems to not need.

Reply:
I changed to return the value which `yield` returns because of following the default behavior. 
https://github.com/fluent/fluentd/blob/9dcf9488c3df269db37917ff09b7b91860d46fac/lib/fluent/plugin/out_forward.rb#L822-L834

https://github.com/fluent/fluentd/pull/2490/commits/60f324947f8d5450011c64ae8f2f73be0ced20fe